query-service-impl/src/main/java/org/hypertrace/core/query/service/projection/ProjectionTransformation.java

```diff
@@ -315,6 +322,77 @@ private Filter rebuildFilterOmittingDefaults(
     return builder.clearChildFilter().addAllChildFilter(childFilters).build();
   }
 
+  private QueryRequest rebuildRequest(
+      QueryRequest original,
+      List<Expression> selections,
+      List<Expression> aggregations,
+      Filter filter,
+      List<Expression> groupBys,
+      List<OrderByExpression> orderBys) {
+
+    Set<AttributeExpression> attributeExpressionSet =
+        Stream.concat(
+                findComplexAttributeExpressionFromOrderBy(orderBys),
+                findComplexAttributeExpressionFromFilter(filter))
+            .collect(Collectors.toSet());
+
+    return rebuildRequestOmittingDefaults(
+        original,
+        selections,
+        aggregations,
+        createFilterForComplexAttributeExpression(attributeExpressionSet, filter),
+        groupBys,
+        orderBys);
+  }
+
+  private Stream<AttributeExpression> findComplexAttributeExpressionFromOrderBy(
+      List<OrderByExpression> orderByExpressionList) {
+    Stream.Builder<AttributeExpression> builder = Stream.builder();
+    for (OrderByExpression orderByExpression : orderByExpressionList) {
+      if (orderByExpression.getExpression().getValueCase() == ValueCase.ATTRIBUTE_EXPRESSION
+          && orderByExpression.getExpression().getAttributeExpression().hasSubpath()) {
+        builder.add(orderByExpression.getExpression().getAttributeExpression());
+      }
+    }
+    return builder.build();
+  }
+
+  private Stream<AttributeExpression> findComplexAttributeExpressionFromFilter(Filter filter) {
+    Stream.Builder<AttributeExpression> builder = Stream.builder();
+
+    for (Filter childFilter : filter.getChildFilterList()) {
+      findComplexAttributeExpressionFromFilter(childFilter).forEach(builder::add);
+    }
+
+    if (filter.getLhs().getValueCase() == ValueCase.ATTRIBUTE_EXPRESSION
+        && filter.getLhs().getAttributeExpression().hasSubpath()) {
+      builder.add(filter.getLhs().getAttributeExpression());
+    }
+    return builder.build();
+  }
+
+  private Filter createFilterForComplexAttributeExpression(
+      Set<AttributeExpression> attributeExpressionSet, Filter filter) {
+
+    if (attributeExpressionSet.isEmpty()) {
+      return filter;
+    }
+
+    List<Filter> childFilterList = new ArrayList<>();
+    childFilterList.add(filter);
+
+    attributeExpressionSet.forEach(
+        attributeExpression ->
+            childFilterList.add(
+                Filter.newBuilder()
+                    .setOperator(Operator.CONTAINS_KEY)
+                    .setLhs(createColumnExpression(attributeExpression.getAttributeId()))
+                    .setRhs(createColumnExpression(attributeExpression.getSubpath()))
+                    .build()));
+
+    return Filter.newBuilder().setOperator(Operator.AND).addAllChildFilter(childFilterList).build();
```

Comment:
@kotharironak 
Is there any problem if there is only one filter in childFilterList?

Reply:
There shouldn't be, but there will always be more than one here. We know attributeExpressionSet is not empty (line 377) and we add a filter on 382, so this will never return more less than two filters.

Reply:
So the behavior we discussed on slack is a bit different. The order by expressions need to be added as filter at the top level like this, but attributes referenced in filters should be anded with that filter rather than promoted.

Reply:
resolved